deps/chakrashim/src/v8string.cc

```diff
@@ -110,15 +110,61 @@ int String::WriteOneByte(
 
 int String::WriteUtf8(
     char *buffer, int length, int *nchars_ref, int options) const {
+  size_t count = 0;
   if (length < 0) {
     // in case length was not provided we want to copy the whole string
     length = String::kMaxLength;
+  } else if (buffer != nullptr) {
+    JsCopyString((JsValueRef)this, nullptr, 0, &count);
+
+    if (length < count || (length == count && !(options & String::NO_NULL_TERMINATION))) {
+      // if length == count there is no space for null terminator
+      // Slow path: must implement truncation here.
+      char* fullBuffer = (char *)malloc(count);
+      CHAKRA_VERIFY(fullBuffer != nullptr);
+
+      JsCopyString((JsValueRef)this, fullBuffer, count, nullptr);
+      memmove(buffer, fullBuffer, sizeof(char) * length);
+      free(fullBuffer);
+
+      // Truncate string to the start of the last codepoint
+      if (length > 0 && ((buffer[length-1] & 0x80) == 0)) {
+        if (!(options & String::NO_NULL_TERMINATION)) {
+          // We must reserve space for a null terminator
+          // We can just remove this one-byte codepoint.
+          length -= 1;
+        }
+      } else if (length > 0 && (buffer[length-1] & 0xC0) == 0xC0) {
```

Comment:
It might make this code more clear to make these magic numbers variables or `#define`s

Reply:
The `0xC0` I converted to a macro since it was used a bunch. I also added in some new checks, but used in only one place so I documented those with comments.